### ISSUE TITLE
Complex query criteria ($all + $elemMatch) doesn't work

### ISSUE THREAD

**Issue by piotrrzacki** [bwaldvogel/mongo-java-server#97]
Dear Benedikt.
I tried to find document using complex query criteria ($all + $elemMatch), but it is not working on last snapshot version.

Simple project to reproduce this bug (I checked this complex query on mongo 3.6.13 and it works):

**build.gradle:**
```
plugins {
    id 'org.springframework.boot' version '2.1.8.RELEASE'
    id 'io.spring.dependency-management' version '1.0.8.RELEASE'
    id 'java'
}

group = 'example'
version = '0.0.1-SNAPSHOT'
sourceCompatibility = '1.8'

repositories {
    mavenLocal()
    mavenCentral()
}

ext {
    set('springBootVersion', "2.1.8.RELEASE")
}

dependencies {
    implementation 'de.bwaldvogel:mongo-java-server:1.21.0-SNAPSHOT:all'
    implementation "org.springframework.boot:spring-boot-starter-data-mongodb"
    implementation "org.springframework.boot:spring-boot-starter-test"
}

dependencyManagement {
    imports {
        mavenBom "org.springframework.boot:spring-boot-starter-parent:${springBootVersion}"
    }
}
```

**example.DemoApplication.java:**
```java
@SpringBootApplication
public class DemoApplication {

    public static void main(String[] args) {
        SpringApplication.run(DemoApplication.class, args);
    }

    @Bean
    public MongoTemplate mongoTemplate(MongoClient mongoClient) {
        return new MongoTemplate(mongoDbFactory(mongoClient));
    }

    @Bean
    public MongoDbFactory mongoDbFactory(MongoClient mongoClient) {
        return new SimpleMongoDbFactory(mongoClient, "test");
    }

    @Bean(destroyMethod="shutdown")
    public MongoServer mongoServer() {
        MongoServer mongoServer = new MongoServer(new MemoryBackend());
        mongoServer.bind();
        return mongoServer;
    }

    @Bean(destroyMethod="close")
    public MongoClient mongoClient(MongoServer mongoServer) {
        return new MongoClient(new ServerAddress(mongoServer.getLocalAddress()));
    }
}
```

**example.DemoApplicationTests.java:**
```java
@RunWith(SpringRunner.class)
@SpringBootTest
public class DemoApplicationTests {

    @Autowired
    MongoTemplate mongoTemplate;

    @Before
    public void before() {
        mongoTemplate.remove(new BasicQuery("{}"), TestDocument.class);
    }

    @Test
    public void contextLoads() {
        // insert document with field as list
        TestDocument doc = new TestDocument(
            UUID.randomUUID().toString(),
            Arrays.asList(
                new InnerDocument("aa", "bb"),
                new InnerDocument("cc", "dd")));
        mongoTemplate.insert(doc, "testDocument");

        TestDocument dbDoc1 = mongoTemplate.findOne(new BasicQuery("{\"list\": {\"$all\": [{\"$elemMatch\": {\"key\": \"aa\", \"value\": \"bb\"} }, {\"$elemMatch\": {\"key\": \"cc\", \"value\": \"dd\"} } ], \"$size\": 2} }"), TestDocument.class);
        TestDocument dbDoc2 = mongoTemplate.findOne(new BasicQuery("{\"list\": {\"$size\": 2, \"$all\": [{\"$elemMatch\": {\"key\": \"aa\", \"value\": \"bb\"} }, {\"$elemMatch\": {\"key\": \"cc\", \"value\": \"dd\"} } ]} }"), TestDocument.class);

        Assertions.assertThat(dbDoc1).isNotNull();
        Assertions.assertThat(dbDoc2).isNotNull();
    }

    @Document
    public static class TestDocument {
        @Id String id;
        List<InnerDocument> list;

        public TestDocument(String id, List<InnerDocument> list) {
            this.id = id;
            this.list = list;
        }
    }

    @Document
    public static class InnerDocument {
        @Id String key;
        String value;

        public InnerDocument(String key, String value) {
            this.key = key;
            this.value = value;
        }
    }

}
```

**Comment by bwaldvogel** [comment]
Can you re-test on c8e9c5f30968a4703fd7e423a5b22bcf835ce923?

**Comment by piotrrzacki** [comment]
It works. Thank you

**Comment by bwaldvogel** [comment]
The fix is part of version `1.21.0`.
